### DISCUSSION TITLE
Feasibility of getting ruby debug gem working with mruby

### DISCUSSION THREAD

**Discussion by jabamaus** [mruby/mruby#5956] [General]
Hi,

I've ported my codebase to use mruby, which I'm loving. However, I do badly miss the ruby debug gem running under vscode. I was thinking of making my codebase work with C ruby too just for this purpose but I really don't want to and feels really nasty. I was wondering, in principle, if its possible to port the gem to mruby, and if so how big a task that might be. I'm assuming the vscode extension could stay the same (I hope). I did mention this in the debug gem discussions but its not on their radar.

Thanks :)

**Comment by matz** [comment]
As far as I understand, the current `debug.gem` depends on tracepoint feature which mruby does not implement. So its porting must be difficult. In general, gems can be ported unless:

* they use C extensions
* they use CRuby specific features
* they hit mruby incompatibility

The biggest obstacle is `require`, which may or may not be solved by `mruby-require` mrbgem.
The last one should be fixed, so feel free to report the issue here.

**Comment by jabamaus** [comment]
Hi Matz. Thanks for replying. Is there any appetite from you guys to implement the tracepoint api in debug build? If not I suppose its dead in the water. That would be a shame as it would be a killer feature for mruby.

Why would `require` be such a problem? Sounds minor compared to tracepoint api...

**Reply by matz** [reply]
`require` load C extension dynamic loading, which mruby does not support, mostly due to portability. In addition, CRuby and mruby have different C API.

**Comment by matz** [comment]
tracepoint is a huge API, so it would take time.

**Reply by jabamaus** [reply]
That's not a no. Could be worse 😀 I would love to see it but I know how busy you guys are.

**Reply by matz** [reply]
The other challenge is memory. mruby is originally designed for embedding environment. Since tracepoint should be built in the core, it may consume more memory. It may be unacceptable. We'll see.

**Reply by jabamaus** [reply]
What if its debug build only and wrapped in config ifdefs?

**Reply by jabamaus** [reply]
Is it worth me opening a feature request issue?

**Reply by matz** [reply]
We already have `mrdb` debugger. Perhaps the issues should be to improve it.
